### PR TITLE
Smallab2

### LINKED ISSUE #19: default to logging, user can disable if they want

### PR DESCRIPTION
**Pull request by octopuscabbage**
Fixes #19 #17 #16 #15 #6 

To fix: #7

### CODE REVIEW

**Comment by octopuscabbage** [comment]
okay now #7 is fixed

File: smallab/runner_implementations/abstract_runner.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+import os
+
+import abc
+import typing
+
+from smallab.experiment import Experiment
+
+
+class AbstractRunner(abc.ABC):
```
Comment (akumaraguru) [inline smallab/runner_implementations/abstract_runner.py:12]: This class is hard to parse. Here's a couple suggestions to make it easier for users to understand these method headers:

1. Add docstrings to explain what each arguement represents in more detail
2. For composite typing specifications, especially those that get reused often, it might be better to alias their type specification, e.g.
```
# in file types.py
spec = typing.List[typing.Dict]

# in file abstract_runner.py
from smallab import types
...
    def run(self, specifications_to_run: types.spec, ...
```
It would be more appropriate to use `typing.NewType()` and `typing.TypeVar()`, but looking through the docs for that gives me a headache...

File: smallab/runner_implementations/abstract_runner.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+    @abc.abstractmethod
+    def run(self, specifications_to_run: typing.List[typing.Dict],
+            run_and_save_fn: typing.Callable[[typing.Dict], typing.Union[None, Exception]]):
+        pass
+
+    def get_completed(self):
+        return self.completed_specifications
+
+    def get_failed_specifications(self, ):
```
Comment (akumaraguru) [inline smallab/runner_implementations/abstract_runner.py:30]: extra comma

File: smallab/runner_implementations/joblib_runner.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+from smallab.experiment import Experiment
+from tqdm import tqdm
+
+from smallab.runner_implementations.abstract_runner import AbstractRunner
+from smallab.utilities.tqdm_to_logger import TqdmToLogger
+
+
+class JoblibRunner(AbstractRunner):
+
+    def run(self, specifications_to_run: typing.List[typing.Dict],
```
Comment (akumaraguru) [inline smallab/runner_implementations/joblib_runner.py:14]: Again, type aliasing would make this much more readable.

File: smallab/runner_implementations/main_process_runner.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+import typing
+
+from tqdm import tqdm
+
+from smallab.experiment import Experiment
+from smallab.runner_implementations.abstract_runner import AbstractRunner
+from smallab.utilities.tqdm_to_logger import TqdmToLogger
+
+
+class MainRunner(AbstractRunner):
```
Comment (akumaraguru) [inline smallab/runner_implementations/main_process_runner.py:11]: Might want to specify this is a simple, single-threaded runner.

File: smallab/runner_implementations/multiprocessing_runner.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+from smallab.runner_implementations.abstract_runner import AbstractRunner
+
+def run_dill_encoded(payload):
+    fun, args = dill.loads(payload)
+    return fun(*args)
+
+def apply_async(pool, fun, args):
+    payload = dill.dumps((fun, args))
+    return pool.apply_async(run_dill_encoded, (payload,))
+
```
Comment (akumaraguru) [inline smallab/runner_implementations/multiprocessing_runner.py:16]: spacing

File: smallab/runner_implementations/multiprocessing_runner.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    return result
+
+
+class MultiprocessingRunner(AbstractRunner):
+    def __init__(self,mp_override=None):
+        """
+        A runner which uses a multiprocessing pool to manage specification running
+        :param mp_override: provide multiprocessing library that should be used. This is done because pytorch has a funky multiprocessing library that could be pased in here
+        """
+        #import multiprocessing_logging
```
Comment (akumaraguru) [inline smallab/runner_implementations/multiprocessing_runner.py:50]: Remove comments

File: setup.py [hunk 7]
```diff
@@ -14,7 +14,7 @@ def read(fname):
 
 setup(
     name = "smallab",
-    version = "0.2.0",
+    version = "2.0.0",
```
[the remark this answers is not in the document]
Comment (octopuscabbage) [reply setup.py:17]: i mean we use it "in production" and this contains a ton of not backward compatible changes so it should have it's major number changed by semvar

[the remark this answers is not in the document]
Comment (octopuscabbage) [reply setup.py:17]: I think this could be 1 and assume the old version is 0 but i don't like 0 indexing in version numbers

[the remark this answers is not in the document]
Comment (akumaraguru) [reply setup.py:17]: That's fair, I'm fine with 1.0.0

File: smallab/runner_implementations/abstract_runner.py (hunk 1 above)
Comment (octopuscabbage) [reply smallab/runner_implementations/abstract_runner.py:12]: I don't think we should have a composite type representing a list of specifications but i will make one for a specification.

File: smallab/runner_implementations/abstract_runner.py (hunk 2 above)
Comment (octopuscabbage) [reply smallab/runner_implementations/abstract_runner.py:30]: im surprised the python compiled with that
